networkx/relabel.py

```diff
@@ -116,44 +118,38 @@ def relabel_nodes(G, mapping, copy=True):
     if not hasattr(mapping, "__getitem__"):
         m = {n: mapping(n) for n in G}
     else:
-        m = mapping
+        # enforce a complete mapping (if only partial mapping is defined)
+        m = {n: mapping.get(n, n) for n in G}
     if copy:
         return _relabel_copy(G, m)
     else:
         return _relabel_inplace(G, m)
 
 
 def _relabel_inplace(G, mapping):
-    old_labels = set(mapping.keys())
-    new_labels = set(mapping.values())
-    if len(old_labels & new_labels) > 0:
+    if len(mapping.keys() & mapping.values()) > 0:
         # labels sets overlap
         # can we topological sort and still do the relabeling?
         D = nx.DiGraph(list(mapping.items()))
         D.remove_edges_from(nx.selfloop_edges(D))
         try:
-            nodes = reversed(list(nx.topological_sort(D)))
+            # only keep the nodes in the original graph
+            nodes = [n for n in nx.topological_sort(D) if n in mapping.keys()]
         except nx.NetworkXUnfeasible as err:
             raise nx.NetworkXUnfeasible(
                 "The node label sets are overlapping and no ordering can "
                 "resolve the mapping. Use copy=True."
             ) from err
     else:
         # non-overlapping label sets
-        nodes = old_labels
+        nodes = mapping.keys()
 
     multigraph = G.is_multigraph()
     directed = G.is_directed()
 
     for old in nodes:
-        # Test that old is in both mapping and G, otherwise ignore.
-        try:
-            new = mapping[old]
-            G.add_node(new, **G.nodes[old])
-        except KeyError:
-            continue
-        if new == old:
-            continue
+        new = mapping.get(old, old)
+        G.add_node(new, **G.nodes.get(old, {}))
```

Comment:
This change should also be backed out to make it work with overlapping node sets.